CBRN Gear Rework: Basic Gasmask

DoctorBoomstick opened the pull request:
#### Summary
Content " Updates the basic gas mask, related filtr, filter ammunition, and gas mask bag to modern standards of CDDA, adds possible inserts and outserts, disassembly craft, and item group goodness."
#### Purpose of change
In pull request #66424, I attempted, unsuccessfully, to update in-game CBRN (chemical, biological, radiological, and nuclear) equipment. The reasons for my failure in this regard are wide and manifold, but suffice it to say I had bitten off more than I could chew in a single PR, and we couldn’t, and still can’t, simulate individual pieces of MOPP gear in any appreciable way. However, because I am a determined idiot and dislike letting dead PRs lie, I am resurrecting, with additions and cleanups, the only part of that PR that was worth extracting: gas mask updates.
- At the present, our in-game CBRN gas masks are showing their age: I can smell the mothballs even stronger than in my closet of cold war nonsense. They are, indeed, antiquated to the point where the gas mask doesn’t even have a defined length and doesn’t make use of certain advances in our armour system.
- The gas mask’s magazine was, sensibly enough, named the gas mask filter. Less sensible, however, is the fact that the ammunition (used here as a derogatory term, for it’s the item that depletes while the player is in contaminated areas) was also called a gas mask filter. This means that you had a gasmask loaded with a gas mask filter, loaded with a gas mask filter. This is a confusing level of name assimilation.
- Item groups were a tad suspect, for in practical CBRN operations as well as regular military and chemical-protection conduct, one is always carrying around spare filters, attachments to enhance the utility of one’s mask, etc. We did not have this, as far as I was able to observe, in any form.

There were many such anomalies along the way, most of which I probably missed berried in some tucked-away file or couldn’t resolve, but I have tried my best to improve upon the gas mask because, hey, we all have our interests, and I just happen to be a strange goober who’s interested in peace’s of rubber and metal you clamp over your face to keep from breathing in nerve agents. I need to get out, touch some grass, and actually see the sun, but in lieu of that, the changes are as follows:.
#### Describe the solution
#### Describe alternatives you've considered
Firstly, it must be noted that I have only touched the basic gas mask item, it’s filter, filter ammo, and carrying bag. I have not touched the XL, XS, or half-face gas masks or crafting recipes, for those are a particular kettle of fish to unpack, and I’ve aimed to keep these PRs manageable in size. They will be touched in due course.
I have:
- Updated the dimensions and statistics of the gas mask across the board.
- Updated the statistics and names of the (newly dubbed) CBRN filter canister, and it’s ammo, now called activated carbon.
- Updated the gasmask carry bag to actually simulate the item it’s comment linked to, rather than linking to it and taking “creative liberties.”.
- Created a number of attachments one can insert into their mask, including sun-lite, breacher, and clear lenses, as well as a vision correction assembly because, yes, you can no longer wear shades with your mask. I don’t actually know if you could before, but you certainly can’t now.
- Created what I deem to be reasonable and realistic packed gas mask bag item groups and set their spawn chances as appropriate within monster drops and other locations where gas masks feature. While I might have missed some, the principle portion of omissions to this update, such as the army base warehouse and mass grave spawns, have been deliberately left untouched to facilitate future plans.
- Added gas mask canister filters to hardware stores and general hardware material spawns as filters, and, indeed, gas masks themselves find use outside of simply military and police uses.
- And a wide assortment of other small tweaks: gas masks are no longer uncomfortable items to wear, they have a disassembly craft, they cannot be used with regular glasses, they can now be poorly fit, etc.
#### Describe alternatives you've considered
Quite a lot, so buckle up; this is going to be a ride in the realms of big plans and broken dreams.
#### Known Issues and Things I couldn’t Resolve
Relatively small issues that were, mostly, already present before my changes but that merit mention:
- Gas masks still allow you to eat, drink, chew on a leg of lamb, smoke cigarettes, play the harmonica, vomit without choking, sip whisky, snack on pain killers, blow whistles, vomit, sip Wolf Boy mutagen, and probably a massive host of other jaw-related motions not possible while having a genuine gas mask on. This, in my opinion, is one of the largest issues plaguing masks, for while some actions are reasonable, such as drinking if you have an appropriate canteen adaptor and a mask with a built-in drinking straw, one should not be allowed to snack on a bag of popcorn while walking around with a respirator. There is, out of C++ wizardry, no solution that I know of, so I didn’t try to fix this particular issue.
- Issue fixed as of the 4th of June. Consult comments and commit history for further details. Gas masks are incompatible with helmets and other head equipment. I attempted to make our masks compatible with helmets, as most contemporary masks allow their use with protective headwear as a matter of convenience and safety; however, I was unable to accomplish this. Despite setting the conflicting regions of the mask to skin-tight as sensible, helmets still conflict with them in my testing, while making them soft items allowed all manner of nonsensical items to be worn along side, such as glasses. I deemed the latter to be a greater evil than the former, so I have elected to leave our masks in mutual rivalry with the helmits.
- Scaling encumbrance on the mouth region of the mask go up as you fill the pockets dedicated to inserts. To make our masks more realistic, I implemented them with the capability to increase mouth encumbrance if the filter pocket’s filled with a filter canister (this sounds fancy, but all I did was make the mouth area of the mask have scaling pocket encumbrance). Fortunately, this works. Unfortunately, this encumbrance also takes into account if you slide an insert behind your mask’s visor, because they’re also considered pockets.

#### Things I Planned on Implementing, but Couldn’t’ or Decided to Postpone
- I wished to, like ballistic plates, build our masks with the weakness that a relatively strong hit will crack the mask, turning it into a breached gas mask and making it mostly useless for filter protection, even with a canister installed. In real life, any damage to a gas mask or the compromising of its seal is enough to render your protection void. Giving this weakness to our in-game masks would lead to them being more fragile pieces of equipment that you had to look out for and couldn’t just walk around with all day every day and use them as glorified face armour. Unfortunately, though I “coded” (as much as I can be said to code when being a JSON chimp) the same break function as is present on our armour plates, no amount of whacking, beating, or trauma of all stripes could prompt the mask to change and become the broken version. I tested with a wide assortment of scenarios, letting several zombies wale away on my character for 2 hours, smashing the gas mask into walls, surrounding myself with machinegun turrets, and letting them blast away for another 2 hours. The mask itself would grow damaged but not provoke its transformation into its compromised state.
- I wished to make filters, once removed from their packaging, start expiring right away, as they would in real life as they filter anything and everything passing through the air. My original plan was to create a “packed filter” item that the player would activate to unseal, providing them with the current filter magazine that would possess a several hour timer until it transformed into an “expired filter” item. I decided to not attempt this plan, as, considering that our filters use the magazine code to actually function, I feared that adding transform conditions would cause bugged behaviour.
- I wanted to make gas masks muffle the player’s voice when speaking or yelling with them equipped. While I believe that this could be accomplished through eoc functionality, I decided to leave this field untapped as it would be strange if only gas masks muffled the player character’s voice while scarves and other mouth coverings don’t, and I only wished to touch one gas mask at a time. I’ll probably try to tackle this in the future, though, outside of linking it to some manner of flag, I am unsure how I’d go about this.
- I considered adding variants to the gas mask, considering that we have historic variants for the fragmentation steel helmet; however, I feel like we have enough issues of confusion in the mix already with 50,000 different gun names to make it a non-prudent move on my behalf to amble in and add: C50 gas mask, FM53 gas mask, CM8 gas mask, CM6 gas mask, MP5 gas mask, etc. I will probably add old versions of gas masks to spawn within pawn and military surplus stores that lack much of the modern functionality of more contemporary respirators, but this has to be done once the current issues are addressed.
- I wished to add different filter types that would protect against different forms of contamination: riot filters to protect against crowd-control riot agents; particle filters to protect only against particulate threats and not vapour hazards; CBRN filters to protect against basically anything under the sun, etc. The filtration code itself is, far as I’m concerned, lord of the rings-level sorcery, and I couldn’t even start to understand how one would filter based on the field the player is standing within.
- I will add craft recipes in the future; however, I have chosen not to add them as of yet to prevent one mask being dramatically more complicated than another to craft while I’m still going through one at a time.
- 
#### Final possible concerns
- The volume for the gas mask is, as best I could tell, the most accurate statistic I could assign to it, which doesn’t say a lot. Gas masks are naturally complicated items to calculate the volume for, and there’s only so much that a blind JSON chimp, such as myself, can do when I have to resort to measuring one of my masks, shoeing the tape measure to an AI image recognizer, and hoping it can read the right numbers I’m showing it.
- I’m not confident that the encumbrance number accurately portrays the effort it takes to do labour with a gas mask. I believe it’s a reasonable representation, but it might be liable for further upscaling if deemed necessary.
- 
#### Testing
- Step one: I occupied a cabin. Was it Liam’s home? Not anymore. Now, it’s the experiment chamber.
- Step 2: I tested out various scenarios of gas exposure, bloater and tearjerker zombies surrounding my character, detonating fungicidal and teargas grenades in my vicinity (mainly in my own hands), and exposure to fungal enemies during different states of readiness: mask off, mask on but not pressurised, mask on and pressurised, and mask on with no canister.
- Step 3: I tested inserting and removing the various inserts and outserts from the mask and observed their effects in all cases. With the vision correction assembly, I mutated nearsightedness and observed that the insert did fix vision problems when installed.
- Step 4: I tested different articles of clothing with and without the mask on to observe what it can and cannot be worn with, and I satisfied myself that skin-tight items, such as glasses and ski masks, cannot be worn.
- Step 5: I tested the encumbrance of the mask at different readiness states: mask on with no filter, mask on with filter attached but not pressurised, mask on with filter attached and pressurised, and mask off, as I sprinted around the room and observed the different rates at which stamina dropped and returned.
- Step 6: I summoned literal armies of feral soldiers to my Kastle (a comfortable CPU-strangling amount of over 1000), killed them all with the might of debug godhood (RIP feral soldiers), and picked through the mountains of trash, bodies, and other stuff to observe gas mask spawns. In this, I noted that MREs, of all items, had a tendency to slip into the gas mask packs that the soldiers were carrying, but I did not notice any other item exhibiting this behaviour.
- 
#### Additional context

DoctorBoomstick committed:
fixes merge errors.

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge branch 'master' of https://github.com/DoctorBoomstick/Cataclysm-DDA

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
updates the filter.

DoctorBoomstick committed:
updates the mask.

DoctorBoomstick committed:
keeps updating the mask.

DoctorBoomstick committed:
adds outserts.

DoctorBoomstick committed:
finishes updating outserts.

DoctorBoomstick committed:
adds deconstruct craft for the gas mask.

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge branch 'master' into gasmask

DoctorBoomstick committed:
ques up a fresh batch of changes for testing.

DoctorBoomstick committed:
bug fixes.

DoctorBoomstick committed:
more bug squishing.

DoctorBoomstick committed:
experement in adding softness to gas masks and seing what happens.

DoctorBoomstick committed:
I am stupid. this adds the soft flag.

DoctorBoomstick committed:
adds spawn locations as required.

DoctorBoomstick committed:
bug fixes and item expansions.

DoctorBoomstick committed:
final few bug fixes.

DoctorBoomstick committed:
final update, in theory, I'm a lier, adjusting material layers after some testing.

anothersimulacrum commented on data/json/itemgroups/Clothing_Gear/gear.json:
```diff
@@ -1,4 +1,77 @@
+    "type": "item_group",
+    "subtype": "collection",
+    "//": "This is a stowed-away gas mask in its carrier, packed with all its equipment but unused and unassembled. Groups together a gas mask, a spare filter, and loose accessories.",
+    "//2": "The gas mask doesn’t have its own filter installed, so a second one is sitting ready in the pouch for the user to assemble.",
```
```suggestion
    "//2": "The gas mask doesn’t have its own filter installed, so a second one is sitting ready in the pouch for the user to assemble.",
```

anothersimulacrum commented on data/json/flags.json:
```diff
@@ -2333,6 +2333,18 @@
+  {
+    "id": "PANORAMIC_OUTSERT",
+    "type": "json_flag",
+    "info": "This item is meant to be externally attached to the eyepiece of a single-Lenz gas mask.",
```
Unless this has some other meaning?
```suggestion
    "info": "This item is meant to be externally attached to the eyepiece of a single-lens gas mask.",
```

anothersimulacrum commented on data/json/flags.json:
```diff
@@ -2333,6 +2333,18 @@
+  {
+    "id": "PANORAMIC_INSERT",
+    "type": "json_flag",
+    "info": "This item is intended to be inserted behind the visor of a single-Lenz gas mask.",
```
```suggestion
    "info": "This item is intended to be inserted behind the visor of a single-lens gas mask.",
```

anothersimulacrum commented on data/json/items/tool_armor.json:
```diff
@@ -1961,37 +1961,112 @@
-    "volume": "1 L",
-    "price": "49 USD",
+    "name": { "str": "gas mask" },
+    "description": "A full-face, industrial-grade respirator made from a solid black silicon and butyl rubber blend, providing decent user comfort in addition to a clear field of view and a secure fit thanks to the mask’s single-element vizor and 6-point head harness.  Provided that a filter canister’s been screwed into the right-side filter port and the respirator’s been prepared, it’ll do an ad ...
```
visor is a more common spelling.
```suggestion
    "description": "A full-face, industrial-grade respirator made from a solid black silicon and butyl rubber blend, providing decent user comfort in addition to a clear field of view and a secure fit thanks to the mask’s single-element visor and 6-point head harness.  Provided that a filter canister’s been screwed into the right-side filter port and the respirator’s been prepared, it’ll do an admirable job at preventing smoke, gas, and contaminants derived from chemical and lab work from taking up residence in your organs.",
```

DoctorBoomstick committed:
Update data/json/itemgroups/Clothing_Gear/gear.json

Co-authored-by: anothersimulacrum <anothersimulacrum@gmail.com>

DoctorBoomstick committed:
Update data/json/flags.json

Co-authored-by: anothersimulacrum <anothersimulacrum@gmail.com>

DoctorBoomstick committed:
Update data/json/flags.json

Co-authored-by: anothersimulacrum <anothersimulacrum@gmail.com>

DoctorBoomstick committed:
Update data/json/items/tool_armor.json

Co-authored-by: anothersimulacrum <anothersimulacrum@gmail.com>

fairyarmadillo commented:
I don't think merging this while the helmet conflict exists is a good idea. Players need both helmets and gas masks.

DoctorBoomstick committed:
Update data/json/itemgroups/corpses.json

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

DoctorBoomstick commented:
I’m unsure why there is even a helmet conflict to start with. All the head-covering parts are on the skin-tight layer, while things like army helmets cover the normal layer. Offhandedly, I believe the gas mask at present is also on the normal layer, so why they don’t conflict is beyond me. That being said, I seem to recall that, at present, it only covers the eyes and mouth and not the head, so that's likely why. still confused as to why they're clashing right now, in this PR, though.

fairyarmadillo commented:
Could be something to do with subparts not being correctly checked on the head. Mouth and eyes predate other subparts, and could be excluded from the usual checks in outfit.cpp.

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Some testing science to see if material shenanigans fix the helmet layering issue.

DoctorBoomstick committed:
fixes the fix, cause I'm speshal that way. should fix helmit issues, testing in progress.

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge branch 'master' into gasmask

fixes merge errors.

DoctorBoomstick committed:
replaces ’ marks with normal ' marks, because github bot doesn't like just sitting and eating it's ’ signs like a good little non-bodied pile of code, pain and shattering dreams.

DoctorBoomstick committed:
replacing more ’ with '. I hope I didn't miss any. God, I hope I didn't miss any.

DoctorBoomstick commented:
Update regarding the helmet conflict issue: I’ve added a fresh rubber material that is classified as soft and added it to the mask’s material list in place of the regular rubber. Helmets, at least the army helmet, now fit; glasses don’t, but you can still pull on ski masks and dust masks because layer conflicts are a lie, I guess. We can just handwave away that by believing that the survivor is just strapping those items over the mask itself, however, so I consider this blocking issue to be solved.

DoctorBoomstick committed:
more ’ murder. life's a lie. I hate this simble. who invented it? why did you invent it? I hate you, like this simble, that the bot hates, and that I hate, the bot. I hate the bot. And this simble, for being hated by the bot, which I hate.

DoctorBoomstick committed:
item weight magic number punching for the outserts, and material updates for the outserts and gas mask considering I just descovered we actualy have impact-resistant plastic as an individual material.

DoctorBoomstick committed:
updates material field for the filter port, and updates the density of hard plastic. online saurses for the material peg it at 1.2, but in game it was only 1. this hopefuly fixes the density tests.

DoctorBoomstick committed:
tries to fix the last density related issues, happens to also fix the bad volumes and mass left over with the vision correction insert.

DoctorBoomstick committed:
tries to fix the dencity, again.

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge branch 'master' into gasmask

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge branch 'master' into gasmask

DoctorBoomstick committed:
Merge remote-tracking branch 'upstream/master'

DoctorBoomstick committed:
Merge branch 'master' into gasmask
